### ISSUE TITLE
dependency() object returns true but library does not exist

### ISSUE THREAD

**Issue by falstaff84** [mesonbuild/meson#2643]
systemd checks for a library like this:

    libqrencode = dependency('libqrencode',
                             required : want_qrencode == 'true')
    have = libqrencode.found()

On a minimal Arch Linux system this does not show anything on the console ("Dependency libqrencode found:"). The library is not installed, however, "have" is true!

When setting "method : 'pkg-config',", the check works.

Diffing the whole meson output between the minimal system and my also Arch Linux based development system actual system revels several missing dependency checks in the output...

Presumably, the minimal system is missing something which breaks the meson dependency check when using auto...

This appeared during investigation/tests around this issue:
https://github.com/systemd/systemd/issues/7367

**Comment by nirbheek** [comment]
I can't reproduce this at all. What are the steps for reproducing this?

**Comment by keszybz** [comment]
I can reproduce this with systemd git checkout:
```console
$ ninja -C build
ninja: Entering directory `build'
[0/1] Regenerating build files.
The Meson build system
Version: 0.43.0
Source dir: /home/fedora/src/systemd
Build dir: /home/fedora/src/systemd/build
Build type: native build
Project name: systemd
Native C compiler: cc (gcc 7.2.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Program tools/meson-check-compilation.sh found: YES (/home/fedora/src/systemd/tools/meson-check-compilation.sh)
Program c++ found: YES (/usr/lib64/ccache/c++)
Native C++ compiler: ccache c++ (gcc 7.2.1)
Compiler for C supports argument -Wextra: YES
...
Message: maximum system UID is 999
Message: maximum system GID is 999
Library rt found: YES
Library m found: YES
Library dl found: YES
Library crypt found: YES
Found pkg-config: /usr/bin/pkg-config (1.3.10)
Dependency libapparmor found: NO
Dependency polkit-gobject-1 found: NO
Library acl found: YES
Library pam found: YES
Library pam_misc found: YES
Library gcrypt found: YES
Library gpg-error found: YES
Library bz2 found: YES
Configuring config.h using configuration
```

while on a different machine I get:
```console
+ meson /root/build -D sysvinit-path=/etc/init.d -D default-hierarchy=unified -D man=false
The Meson build system
Version: 0.43.0
Source dir: /root/src
Build dir: /root/build
Build type: native build
Project name: systemd
Native C compiler: cc (gcc 7.2.0)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Message: Activated pre-commit hook
Program tools/meson-check-compilation.sh found: YES (/root/src/tools/meson-check-compilation.sh)
Program c++ found: YES (/usr/sbin/c++)
Native C++ compiler: c++ (gcc 7.2.0)
Compiler for C supports argument -Wextra: YES
...
Message: maximum system UID is 999
Message: maximum system GID is 999
Dependency threads found: YES
Library rt found: YES
Library m found: YES
Library dl found: YES
Library crypt found: YES
Found pkg-config: /usr/sbin/pkg-config (0.29.2)
Native dependency libcap found: YES 2.25
Native dependency mount found: YES 2.31.0
Native dependency libseccomp found: YES 2.3.2
Dependency libselinux found: NO
Dependency libapparmor found: NO
Dependency polkit-gobject-1 found: NO
Library acl found: YES
Dependency audit found: NO
Native dependency blkid found: YES 2.31.0
Native dependency libkmod found: YES 24
Library pam found: YES
Library pam_misc found: YES
Native dependency libmicrohttpd found: YES 0.9.57
Native dependency libcryptsetup found: YES 1.7.5
Native dependency libcurl found: YES 7.57.0
Native dependency libidn found: YES 1.33
Native dependency libiptc found: YES 1.6.1
Native dependency libqrencode found: YES 4.0.0
Library gcrypt found: YES
Library gpg-error found: YES
Native dependency gnutls found: YES 3.5.16
Native dependency libdw found: YES 0.170
Native dependency zlib found: YES 1.2.11
Library bz2 found: YES
Native dependency liblzma found: YES 5.2.3
Native dependency liblz4 found: YES 1.8.0
Native dependency xkbcommon found: YES 0.7.2
Native dependency glib-2.0 found: YES 2.54.0
Native dependency gobject-2.0 found: YES 2.54.0
Native dependency gio-2.0 found: YES 2.54.0
Native dependency dbus-1 found: YES 1.12.2
Configuring config.h using configuration
```
The first one seems bad, it's not showing the result of various checks, the second one is good.
The first one is F27 the secone one is arch. The first one is a rebuild, the second one is from scratch.

The important part is:
```diff
 Has header "valgrind/valgrind.h": NO
 Message: maximum system UID is 999
 Message: maximum system GID is 999
+Dependency threads found: YES
 Library rt found: YES
 Library m found: YES
 Library dl found: YES
 Library crypt found: YES
-Found pkg-config: /usr/bin/pkg-config (1.3.10)
+Found pkg-config: /usr/sbin/pkg-config (0.29.2)
+Native dependency libcap found: YES 2.25
+Native dependency mount found: YES 2.31.0
+Native dependency libseccomp found: YES 2.3.2
+Dependency libselinux found: NO
 Dependency libapparmor found: NO
 Dependency polkit-gobject-1 found: NO
 Library acl found: YES
+Dependency audit found: NO
+Native dependency blkid found: YES 2.31.0
+Native dependency libkmod found: YES 24
 Library pam found: YES
 Library pam_misc found: YES
+Native dependency libmicrohttpd found: YES 0.9.57
+Native dependency libcryptsetup found: YES 1.7.5
+Native dependency libcurl found: YES 7.57.0
+Native dependency libidn found: YES 1.33
+Native dependency libiptc found: YES 1.6.1
+Native dependency libqrencode found: YES 4.0.0
 Library gcrypt found: YES
 Library gpg-error found: YES
+Native dependency gnutls found: YES 3.5.16
+Native dependency libdw found: YES 0.170
+Native dependency zlib found: YES 1.2.11
 Library bz2 found: YES
+Native dependency liblzma found: YES 5.2.3
+Native dependency liblz4 found: YES 1.8.0
+Native dependency xkbcommon found: YES 0.7.2
+Native dependency glib-2.0 found: YES 2.54.0
+Native dependency gobject-2.0 found: YES 2.54.0
+Native dependency gio-2.0 found: YES 2.54.0
+Native dependency dbus-1 found: YES 1.12.2
 Configuring config.h using configuration
```
(the lines with + are missing from the first log)

**Comment by nirbheek** [comment]
The reason why the `dependency()` lines are missing is because those are cached. We should print `(cached)` after the line instead. I am not sure if this is the same as the original bug; the dependency is found, we just don't print it.

**Comment by keszybz** [comment]
So... not sure if this is a bug or an unexpected feature, but after `touch meson.build && ninja -C build`, the lines for the dependencies tests are not printed again. It's not immediately obvious if the tests themselves are being performed.

It seems that they are not. I can reproduce the issue like this:
```bash
sudo dnf build-dep systemd
git clone https://github.com/systemd/systemd
meson -Dman=false build
ninja -C build
sudo dnf remove  qrencode-devel
touch meson.build
ninja -C build
```
In the reconfiguration phase, qrencode is not rechecked and the build fails with `../src/journal/journal-qrcode.c:22:10: fatal error: qrencode.h: No such file or directory`.

**Comment by keszybz** [comment]
So... I don't think caching like this makes sense. If it looks like stuff is being rechecked, it must be rechecked. Caching results between configure runs breaks everything.

**Comment by nirbheek** [comment]
Ah, if that's the bug then, yes. The reason why found-dependencies are cached is because you might set `PKG_CONFIG_PATH`, etc, during configure and not during the actual build (esp. when generating vcxproj files), so rechecking will mean they won't be found.

We also cannot reliably store those env vars because if those variables are then redefined during the build, did the user mean to re-set that or is that just set unintentionally in the environment and we should ignore it?

I do agree that the way we cache dependencies right now is not great, but we have to come up with a consistent and non-surprising way of dealing with this issue.

**Comment by falstaff84** [comment]
@keszybz with "on a different machine" you mean when using mkosi right? I think mkosi is reusing the build directory of the host. So the problem would come down to that mkosi reuses cached configure data from the host...

**Comment by keszybz** [comment]
I also checked how meson behaves when a dependency was satisfied but stops being satisfied because a higher version is installed:
```python
libqrencode = dependency('libqrencode', version : '<3.4.4', requires : true)
```
This is not handled properly either, because the cached value is used. And while one could say that removing or downgrading of dependencies requires explicit user action, *upgrading* dependencies is something that happens all the time.

**Comment by keszybz** [comment]
> I think mkosi is reusing the build directory of the host.

Yes, exactly. In effect it's the same as uninstalling some packages between two runs.

> The reason why found-dependencies are cached is because you might set PKG_CONFIG_PATH, etc, during configure and not during the actual build

I see. I think that's a work-around for a user bug. Essentially, if one wants to use PKG_CONFIG_PATH, they need to make sure that it is set to that value whenever the configure step is performed.

**Comment by keszybz** [comment]
A related issue, maybe you can enlighten me: what's the procedure to reconfigure anything set from variables, let's say CFLAGS:
```console
$ CFLAGS=... meson configure build
→ does not work, it just prints the configuration
$ CFLAGS=... meson configure build -Dsomething-unrelated=foobar
→ does not work, the new value of $CFLAGS is ignored
```
With autotools I'd say ./configure CFLAGS=... and it'd notice the new value. I hope meson can implement something like this.

**Comment by jpakkane** [comment]
`CFLAGS` et al are converted to Meson options which you can change with `meson configure`.

In Meson we have made a conscious choice to avoid using environment variables as much as possible. Global mutable state is bad enough, but envvars and their setting is, on top of that, mostly invisible and hidden.

**Comment by NickeZ** [comment]
> We also cannot reliably store those env vars because if those variables are then redefined during the build, did the user mean to re-set that or is that just set unintentionally in the environment and we should ignore it?

I think we should ignore environment variables during build and only used cached ones from `meson build` and `meson configure`. I think this will improve reproducibility. IMO a build directory that could build the project at one point should always be able to rebuild the project..

> I also checked how meson behaves when a dependency was satisfied but stops being satisfied because a higher version is installed:

I have a fix that might solve that. Basically it checks the versions for the cached dependencies as well. But I haven't polished it up for merging yet.

https://github.com/mesonbuild/meson/pull/2581

**Comment by nirbheek** [comment]
> I see. I think that's a work-around for a user bug. Essentially, if one wants to use PKG_CONFIG_PATH, they need to make sure that it is set to that value whenever the configure step is performed.

Not necessarily. For instance, with vcxproj files, you configure in `cmd.exe`, and your Visual Studio instance will not have the same variables set. To be fair, the ideal way to solve this would be to always cache environment variables unless the user manually calls `nina reconfigure`. I'd like to hear @jpakkane's thoughts on this.

**Comment by jpakkane** [comment]
I've said it before and I'll say it again: caching environment variables is a disaster area waiting to punch you in the face when you least expect it. It is very difficult to make work and be intuitive to most developers (people with 10+ years of command line dev work with Autotools et al are a minority, not a majority), especially when considering cross platform development.

Printing `(cached)` in the log is good, it makes it clear where the data is coming from.

One thing which was voiced some time in the past was to store the path and timestamp of the `.pc` file where the dependency comes from and rechecking that it is still valid when using a cached dependency. Maybe something like that would be a better solution instead?

**Comment by nirbheek** [comment]
Caching in general is a 'disaster area', especially caching with no obvious way to override it, but we still have to do it. Our dependency caching mechanism is actually already quite complicated and error-prone, so I'm thinking of ways to simplify it. Perhaps refactoring is a better option.

Checking the timestamp of the `.pc` file would solve it for pkg-config dependencies, but we'd need something else for other dependency types.

**Comment by jpakkane** [comment]
If we could never cache any dependency information but instead just always get it from the system it would be totally awesome and would reduce complexity by a fair bit. But is this something we can actually do, especially on Windows?

**Comment by nirbheek** [comment]
Yes, if we cache the environment and print it on reconfigure ;)

**Comment by NickeZ** [comment]
Sorry for hijacking the thread for a question. Would it be correct to use the vs2017 compiler just because you've opened a 2017 dev console?

What would be the correct behavior of `--backend=vs2015` in a 2017 dev console?

**Comment by nirbheek** [comment]
`--backend=vs*` control the format of the vcxproj files, and the current environment and detected compiler are used for running configuration tests. The two are independent right now, but we may decide to change that.
